src/digital/v2.rs

```diff
@@ -136,3 +136,27 @@ pub trait InputPin {
     /// Is the input pin low?
     fn is_low(&self) -> Result<bool, Self::Error>;
 }
+
+/// Describe pin that can switch its input/ouput mode on fly.
+/// Should be used to communicate with custom protocols based hardware.
+///
+/// *This trait is available if embedded-hal is built with the `"unproven"` feature.*
+#[cfg(feature = "unproven")]
+pub trait InputOutputPin<Error>: InputPin<Error = Error> + OutputPin<Error = Error> {
+    /// Switch pin into output mode,
+    /// 
+    /// Result::Ok value is true if pin mode was changed or false if pin already was in output mode.
+    ///
+    /// Expected behaviour after success result:
+    /// - OutputPin functions should work as for ouptut pin
+    /// - InputPin functions should return errors
+    fn mode_output(&mut self) -> Result<bool, Error>;
```

Comment:
These functions will (more most real world uses) require inline blocking due to the read-modify-write operations involved to change between input an output which is generally frowned upon.

Reply:
Yes, indeed, but there are some platform related options available to make it more safe like using asm instructions to block interrupts or using CAS instructions.
I know that it would not be fastest possible implementation, but it has to be as easy to use as possible. Performance drawbacks can be described in trait documentation. BTW there is an opinion that for such cases performance (1 pin bus throughput) should not be big deal.

Reply:
Maybe `fn mode_output(Self) -> Result<Self::Output, Self::Error>`?